File: PyProject/geeksforgeeks/quicksort.py
# -*- coding:utf-8 -*-
# author: youdi

from typing import List


def partition(arr: List[int], low: int, high: int) -> int:
    i = low - 1
    pivot = arr[high]
    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[j], arr[i] = arr[i], arr[j]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


def quicksort(arr: List[int], low: int, high: int):
    if low < high:
        pi = partition(arr, low, high)
        quicksort(arr, low, pi - 1)
        quicksort(arr, pi + 1, high)


if __name__ == '__main__':
    A = [64, 25, 12, 22, 11]
    quicksort(A, 0, len(A) - 1)
    print(A)

File: GoProject/tx/quick_sort.go
package tx

func partition(nums []int, start, end int) int {
	pivot := nums[end]
	i := start - 1

	for j := start; j < end; j++ {
		if nums[j] < pivot {
			i += 1
			nums[j], nums[i] = nums[i], nums[j]
		}

	}

	nums[i+1], nums[end] = nums[end], nums[i+1]
	return i + 1
}

func quick_sort(nums []int, start, end int) {
	if start < end {
		pivot := partition(nums, start, end)

		quick_sort(nums, start, pivot-1)
		quick_sort(nums, pivot+1, end)

	}

}
